From ea7941e23c7c51b313dc6668660c418f0dbe4e3a Mon Sep 17 00:00:00 2001
From: Ray Strode <rstrode@redhat.com>
Date: Thu, 7 Oct 2010 21:47:51 -0400
Subject: [PATCH 1/3] user-manager: fixes for when consolekit is unavailable

Part of the user manager load sequence involves communicating
with ConsoleKit.  If ConsoleKit is unavailable, we need to
properly finish the sequence so that the user manager
becomes accessible in a "degraded" mode.
---
 gui/simple-greeter/gdm-user-manager.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/gui/simple-greeter/gdm-user-manager.c b/gui/simple-greeter/gdm-user-manager.c
index fd9ac99..61c3d7c 100644
--- a/gui/simple-greeter/gdm-user-manager.c
+++ b/gui/simple-greeter/gdm-user-manager.c
@@ -566,6 +566,7 @@ on_get_seat_id_finished (DBusGProxy     *proxy,
                                  "current session");
                 }
                 unload_seat (manager);
+                maybe_set_is_loaded (manager);
                 return;
         }
 
@@ -765,6 +766,7 @@ on_get_current_session_finished (DBusGProxy     *proxy,
                         g_debug ("Failed to identify the current session");
                 }
                 unload_seat (manager);
+                maybe_set_is_loaded (manager);
                 return;
         }
 
-- 
1.7.1


From 16b6f4efe41efa8921f185817a837ae9842040e1 Mon Sep 17 00:00:00 2001
From: Marius Vollmer <mvollmer@redhat.com>
Date: Tue, 13 Aug 2013 17:28:29 -0400
Subject: [PATCH 2/3] Don't set is-loaded prematurely

maybe_set_is_loaded immediately.  This will set "is-loaded" even though
load_users hasn't been called yet.

One path is

load_idle
load_seat_incrementally
get_current_session_id
_get_current_systemd_session_id
unload_seat
maybe_set_is_loaded

This commits fixes it by interchanging the calls to load_seat_incrementally and
load_users in load_idle.
---
 gui/simple-greeter/gdm-user-manager.c |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/gui/simple-greeter/gdm-user-manager.c b/gui/simple-greeter/gdm-user-manager.c
index 61c3d7c..81eb203 100644
--- a/gui/simple-greeter/gdm-user-manager.c
+++ b/gui/simple-greeter/gdm-user-manager.c
@@ -185,6 +185,7 @@ static void     queue_load_seat_and_users   (GdmUserManager *manager);
 static void     monitor_local_users         (GdmUserManager *manager);
 
 static void     load_new_session_incrementally (GdmUserManagerNewSession *new_session);
+static void     maybe_set_is_loaded (GdmUserManager *manager);
 
 static gpointer user_manager_object = NULL;
 
@@ -2108,8 +2109,8 @@ static gboolean
 load_idle (GdmUserManager *manager)
 {
         manager->priv->seat.state = GDM_USER_MANAGER_SEAT_STATE_UNLOADED + 1;
-        load_seat_incrementally (manager);
         load_users (manager);
+        load_seat_incrementally (manager);
         manager->priv->load_id = 0;
 
         return FALSE;
-- 
1.7.1


From ed6fe7fbb7906732f3c1cb4cd3c8db352022f8e2 Mon Sep 17 00:00:00 2001
From: Ray Strode <rstrode@redhat.com>
Date: Tue, 13 Aug 2013 18:16:20 -0400
Subject: [PATCH 3/3] user-manager: better support multiple concurrent queued loads

---
 gui/simple-greeter/gdm-user-manager.c |    5 +++--
 1 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/gui/simple-greeter/gdm-user-manager.c b/gui/simple-greeter/gdm-user-manager.c
index 81eb203..a2eb914 100644
--- a/gui/simple-greeter/gdm-user-manager.c
+++ b/gui/simple-greeter/gdm-user-manager.c
@@ -2077,8 +2077,6 @@ load_sessions (GdmUserManager *manager)
 static void
 load_seat_incrementally (GdmUserManager *manager)
 {
-        g_assert (manager->priv->seat.proxy == NULL);
-
         switch (manager->priv->seat.state) {
         case GDM_USER_MANAGER_SEAT_STATE_GET_SESSION_ID:
                 get_current_session_id (manager);
@@ -2108,6 +2106,9 @@ load_seat_incrementally (GdmUserManager *manager)
 static gboolean
 load_idle (GdmUserManager *manager)
 {
+        if (manager->priv->is_loaded) {
+                return FALSE;
+        }
         manager->priv->seat.state = GDM_USER_MANAGER_SEAT_STATE_UNLOADED + 1;
         load_users (manager);
         load_seat_incrementally (manager);
-- 
1.7.1

